patroni/validator.py

```diff
@@ -0,0 +1,338 @@
+#!/usr/bin/env python3
+import logging
+import os
+import socket
+import re
+import subprocess
+
+from patroni.utils import split_host_port, data_directory_is_empty
+from patroni.ctl import find_executable
+from patroni.dcs import dcs_modules
+from patroni.exceptions import ConfigParseError
+
+logger = logging.getLogger(__name__)
+
+
+def data_directory_empty(data_dir):
+    if os.path.isfile(os.path.join(data_dir, "global", "pg_control")):
+        return False
+    return data_directory_is_empty(data_dir)
+
+
+def validate_connect_address(address):
+    if not isinstance(address, str):
+        raise ConfigParseError("is not a string")
+    else:
+        for s in ["127.0.0.1", "0.0.0.0", "*", "::1"]:
+            if s in address:
+                raise ConfigParseError("must not contain {}".format(s))
+    return True
+
+
+def validate_host_port(host_port, listen=False, connect=False):
+    try:
+        host, port = split_host_port(host_port, None)
+    except (ValueError, TypeError):
+        raise ConfigParseError("contains a wrong value")
+    else:
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            if s.connect_ex((host, port)) == 0 and listen:
+                ConfigParseError("Port {} is already in use.".format(port))
+            elif connect:
+                ConfigParseError("{} is not reachable".format(host_port))
+        except socket.gaierror as e:
+            raise ConfigParseError(e)
+        finally:
+            s.close()
+    return True
+
+
+def comma_separated_host_port(string):
+    if not isinstance(string, str):
+        raise ConfigParseError("is not a string")
+    assert all([validate_host_port(s.strip()) for s in string.split(",")]), "didn't pass the validation"
+    return True
+
+
+def validate_host_port_listen(host_port):
+    return validate_host_port(host_port, listen=True)
+
+
+def is_ipv4_address(ip):
+    assert socket.inet_aton(ip)
+    return True
+
+
+def is_ipv6_address(ip):
+    assert socket.inet_pton(socket.AF_INET6, ip)
+    return True
+
+
+def get_major_version():
+    version = subprocess.check_output(['postgres', '--version']).decode()
+    version = re.match(r'^[^\s]+ [^\s]+ (\d+)(\.(\d+))?', version)
+    return '.'.join([version.group(1), version.group(3)]) if int(version.group(1)) < 10 else version.group(1)
+
+
+def validate_data_dir(data_dir):
+    if not data_dir:
+        raise ConfigParseError("is an empty string")
+    elif os.path.exists(data_dir) and not os.path.isdir(data_dir):
+        raise ConfigParseError("is not a directory")
+    elif not data_directory_empty(data_dir):
+        if not os.path.exists(os.path.join(data_dir, "PG_VERSION")):
+            raise ConfigParseError("doesn't look like a valid data directory")
+        else:
+            with open(os.path.join(data_dir, "PG_VERSION"), "r") as version:
+                pgversion = int(version.read())
+            if str(pgversion) != get_major_version():
```

Comment:
It could be that the postgres isn't available in the default PATH, we have to take postgresql.bin_dir into account.